File: src/main/java/hmm/Pi.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package hmm;

import java.util.*;

/**
 *
 * @author ray
 */
public class Pi {

    private Map<Integer, Double> pi = new HashMap<Integer, Double>();
    private int total = 1;

    public Pi() {
    }

    public Pi(Map<Integer, Integer> pii) {
        Set<Integer> keySet = pii.keySet();
        for (Integer i : keySet) {
            total += pii.get(i);
        }

        for (Integer i : keySet) {
            setPi(i, (double) pii.get(i) / (double) (total));
        }
    }

    public double getPi(int index) {
        if (pi.containsKey(index)) {
            return pi.get(index);
        } else {
            return 1.0 / (double) total;
        }
    }

    public void setPi(int index, double prob) {
        pi.put(index, prob);
    }
}

File: lib.hmm/src/main/scala/hmm/Pi.scala
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package hmm

import scala.collection.mutable.Map

class Pi {

  val pi = Map[Int, Double]();
  var total = 1;

  def getPi(index: Int): Double = pi.getOrElse(index,1.0/total)

  def setPi(index: Int, prob: Double): Unit = pi.put(index,prob)
}

object Pi {
  def apply(pii: Map[Int, Int]): Pi = {
    val keySet = pii.keySet;
    val pi = new Pi()
    var t = 0
    keySet.foreach(t += pii(_))
    pi.total = t
    for (i <- keySet) {
      pi.setPi(i, pii(i).asInstanceOf[Double] / t);
    }
    pi
  }
}